hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala

```diff
@@ -317,22 +317,38 @@ abstract class PContainer extends PIterable {
       return Code._empty
     }
 
-    val i = mb.newLocal[Long]
+    val i = mb.newLocal[Int]
+    val mod = mb.newLocal[Int]
     Code(
-      i := PContainer.nMissingBytes(sourceType.loadLength(sourceOffset)).toL,
-      Code.whileLoop(i > 0L,
-        (i >= 8L).mux(
+      mod := (sourceType.loadLength(sourceOffset) % const(8)) - const(1),
```

Comment:
I think forwards would be much simpler:
1. compute number of elements N
2. compute smallest multiple of 32 less than N, M1 (see below for why not longs)
3. while i < M1, compare ints
4. compute smallest multiple of 8 less than N, M2
5. while i < M2, compare bytes
6. while i < N, compare bits

Reply:
actually you don't even need to compute M1 / M2, you just do `while i + 32 < N`

Reply:
This is doing exactly what you describe, except I need to check how many bits we need to check, else we are going into undefined bits, that may be initialized to 1's

Reply:
I wrote it going backwards because it's marginally more efficient, I think. To figure out how many bytes I can read in their entirety, I need to first look at the top bits. Then, if that wasn't the only byte, I need to subtract one from that byte to get the total number of bytes I can read in their entirety. Then I can choose to continue walking backwards, or go forward. If I walk forward, I start with 1, and need to figure out if I have 7 more bytes left (at least), so that I can read a group of 8. So I need to check if i + 7 <= nMissingBytes. This is a bit slower that just checking i >= 8. I don't know about the cache behavior though, so maybe that offsets it.

Reply:
Tim, just to be sure, w

> Ah, I thought I said I was happy to fix the optimized version rather than revert. I do think it can be simplified, though, per my comments.
> 
> Cotton also had the suggestion of writing this function unstaged using two utility functions:
> 
> ```scala
> def findFirstNonZeroByte(addr: Long, n: Long): Long
> def allPresent(addr: Long, n: Long): Long // uses findFirstNonZeroByte
> ```

How do we convert from address: Code[Long] to Long without going through emit?

Reply:
Not exactly sure how this works. Let's say I have 100 elements.
N = 100 (step 1)
M1 = floor(N / 32) == 3 (step 2)
run 3 iterations, from 0 to 3 exclusive, load int in each loop (step 3)
M2 = floor(N / 8) == 12 (step 4)
I is 3, run 9 iterations checking 3 to 12 exclusive, load byte in each loop (step 5)

By this point I've checked 32 * 3 + 9 * 8 values.

I think I need 3 * 32 + floor((100 - 3*32) / 8), and then change I to be the sum of those, and then check the remaining bits (I in step 6 is in bit terms, not byte, so we have to have a separate counter to track the number of elements checked). So will be more involved.